### PR TITLE
Fixed #29324 -- Made Settings raise ImproperlyConfigured on SECRET_KEY access; not in init.

### PR DESCRIPTION
**Pull request by jdufresne**
https://code.djangoproject.com/ticket/29324

### CODE REVIEW

**Comment by timgraham** [comment]
`SECRET_KEY = ''` is now allowed which I don't think should be the case.

**Comment by jdufresne** [comment]
I see your point. I've updated the change to now do both:

- If `SECRET_KEY` is set and is empty, raise `ImproperlyConfigured` in `__init__`
- If `SECRET_KEY` is not set, do not raise and error until accessed

Thanks!

File: docs/ref/settings.txt [hunk 1]
```diff
@@ -2120,6 +2122,12 @@ affect them.
     startproject <startproject>` creates a unique ``SECRET_KEY`` for
     convenience.
 
+.. versionchanged:: 2.1
+
+   ``SECRET_KEY`` is no longer defined by default. Django will now raise an
```
Comment (timgraham) [inline docs/ref/settings.txt]: I think this is sufficient, "In older versions, SECRET_KEY defaults to an empty string." (use 4 space indent)

File: docs/ref/settings.txt [hunk 2]
```diff
@@ -2087,7 +2087,9 @@ Uses of the key shouldn't assume that it's text or bytes. Every use should go
 through :func:`~django.utils.encoding.force_text` or
 :func:`~django.utils.encoding.force_bytes` to convert it to the desired type.
 
-Django will refuse to start if :setting:`SECRET_KEY` is not set.
+Django will refuse to start if :setting:`SECRET_KEY` is set to an empty value.
+An :class:`~django.core.exceptions.ImproperlyConfigured` exception will be
```
Comment (timgraham) [inline docs/ref/settings.txt]: is raised if SECRET_KEY is accessed but not set.

File: tests/admin_scripts/tests.py [hunk 3]
```diff
@@ -2213,7 +2213,12 @@ def test_unified(self):
         out, err = self.run_manage(args)
         self.assertNoOutput(err)
         self.assertOutput(out, "+ FOO = 'bar'")
-        self.assertOutput(out, "- SECRET_KEY = ''")
+        self.assertOutput(out, "- INSTALLED_APPS = []")
+        self.assertOutput(
+            out,
+            "+ INSTALLED_APPS = ['django.contrib.auth', 'django.contrib.contenttypes', 'admin_scripts']"
+        )
+        self.assertNotInOutput(out, "- SECRET_KEY = ''")
```
Comment (timgraham) [inline tests/admin_scripts/tests.py]: I think this assertion isn't needed.

**Comment by jdufresne** [comment]
Thanks for the review Tim. I've made the suggested changes.

**Pull request merged**